openmc/tallies.py

```diff
@@ -0,0 +1,3204 @@
+
+        """
+
+        cv.check_type('sparse', sparse, bool)
+
+        # Convert NumPy arrays to SciPy sparse LIL matrices
+        if sparse and not self.sparse:
+            if self._sum is not None:
+                self._sum = sps.lil_matrix(self._sum.flatten(), self._sum.shape)
+            if self._sum_sq is not None:
+                self._sum_sq = sps.lil_matrix(self._sum_sq.flatten(),
+                                              self._sum_sq.shape)
+            if self._mean is not None:
+                self._mean = sps.lil_matrix(self._mean.flatten(),
+                                            self._mean.shape)
+            if self._std_dev is not None:
+                self._std_dev = sps.lil_matrix(self._std_dev.flatten(),
+                                               self._std_dev.shape)
+
+            self._sparse = True
+
+        # Convert SciPy sparse LIL matrices to NumPy arrays
+        elif not sparse and self.sparse:
+            if self._sum is not None:
+                self._sum = np.reshape(self._sum.toarray(), self.shape)
+            if self._sum_sq is not None:
+                self._sum_sq = np.reshape(self._sum_sq.toarray(), self.shape)
+            if self._mean is not None:
+                self._mean = np.reshape(self._mean.toarray(), self.shape)
+            if self._std_dev is not None:
+                self._std_dev = np.reshape(self._std_dev.toarray(), self.shape)
+            self._sparse = False
+
+    def remove_score(self, score):
+        """Remove a score from the tally
+
+        Parameters
+        ----------
+        score : str
+            Score to remove
+
+        """
+
+        if score not in self.scores:
+            msg = 'Unable to remove score "{0}" from Tally ID="{1}" since ' \
+                  'the Tally does not contain this score'.format(score, self.id)
+            ValueError(msg)
+
+        self._scores.remove(score)
+
+    def remove_filter(self, old_filter):
+        """Remove a filter from the tally
+
+        Parameters
+        ----------
+        old_filter : openmc.Filter
+            Filter to remove
+
+        """
+
+        if old_filter not in self.filters:
+            msg = 'Unable to remove filter "{0}" from Tally ID="{1}" since the ' \
+                  'Tally does not contain this filter'.format(old_filter, self.id)
+            ValueError(msg)
+
+        self._filters.remove(old_filter)
+
+    def remove_nuclide(self, nuclide):
+        """Remove a nuclide from the tally
+
+        Parameters
+        ----------
+        nuclide : openmc.Nuclide
+            Nuclide to remove
+
+        """
+
+        if nuclide not in self.nuclides:
+            msg = 'Unable to remove nuclide "{0}" from Tally ID="{1}" since the ' \
+                  'Tally does not contain this nuclide'.format(nuclide, self.id)
+            ValueError(msg)
+
+        self._nuclides.remove(nuclide)
+
+    def _can_merge_filters(self, other):
```

Comment:
These `_can_merge_*` blocks contain some complicated logic. It might be prudent to create a base class called `Mergeable` that implements the `can_merge` and `merge` methods so the functionality for `filters`, `nuclides`, and `scrores` could be removed from this file and into the file specifically for the mergeable type. This `Tally` object could also extend from that type. In that way, each individual mergeable could be tested independently.